### PR TITLE
GitHub Actions: update Ubuntu version

### PR DESCRIPTION
**Pull request by cbjeukendrup**
18.04 is deprecated, use 20.04 instead

Resolves:
Scherm­afbeelding 2022-08-22 om 14 39 37

### CODE REVIEW

**Comment by igorkorsukov** [comment]
@cbjeukendrup Maybe better move to 22.04?

**Comment by cbjeukendrup** [comment]
Also fine for me. I chose 20.04 because we seem to have a habit of using the oldest available version (for example, also macOS 11 instead of 12). And `ubuntu-latest` also still points to 20.04.

**Comment by igorkorsukov** [comment]
> 

Ok, I agree with you.

**Review by igorkorsukov — APPROVED** [review]

**Comment by cbjeukendrup** [comment]
Looks like we also have to update GCC version (was 7). Which one should we take?

**Comment by igorkorsukov** [comment]
> Looks like we also have to update GCC version (was 7). Which one should we take?

I believe, we can use default for the given ubuntu version

**Comment by cbjeukendrup** [comment]
Looks like that's 11. So now we have Ubuntu 22.04 with GCC 11.

**Comment by shoogle** [comment]
> Maybe better move to 22.04?

> I chose 20.04 because we seem to have a habit of using the oldest available version

We use the oldest supported version because it gives the widest compatibility when creating AppImages. A Linux binary compiled against one set of libraries will usually run against newer versions of those libraries but not older versions. An AppImage created on 20.04 will run on 22.04 but the other way round won't work.

Ideally we would build binaries on an older distro and run our tests on a newer distro so that we get maximum warning about library API changes, but that takes effort to maintain. If we have to pick one or the other then older is better.

**Review by cr4yfish — APPROVED** [review]
I'm no expert in converting Python2 to 3 but it looks OK.

Also, I agree that choosing Ubuntu 20 would be the right call here.

**Comment by cbjeukendrup** [comment]
Okay, back to Ubuntu 20.

I assume using a newer _compiler_ (GCC 11 instead of for example 9) does not have a negative effect on compatibility?

**Comment by cr4yfish** [comment]
>I assume using a newer compiler (GCC 11 instead of for example 9) does not have a negative effect on compatibility?

Upgrading two full versions will probably introduce some breaking changes - which negatively impact compatibility.
The real question (and the hard one) is: "Are there any changes that affect the current repo?".

I see two ways to figure that out:

1. Change it and search for bugs
2. Do (lots of) research and fix possible bugs beforehand

[Changelog for GCC9 -> GCC10](https://gcc.gnu.org/gcc-10/changes.html)
[Changelog for GCC10 -> GCC11](https://gcc.gnu.org/gcc-11/changes.html)

**Comment by cbjeukendrup** [comment]
(In fact it's not 2 but 4 major versions, since we were using GCC 7. I mentioned 9 because that seems to be the lowest available on Ubuntu 20.04.)

I believe some of us use GCC 11 for local builds, so we do have some experience with it. And there is still enough time to find bugs before the next release... but I don't expect that a different compiler version will cause bugs in our kind or project.

**Comment by cbjeukendrup** [comment]
I think I've found the problem, and it's very weird. If you use `$HOME` directly in the yml file, its value is `/home/runner`, but if you use it in a sh script that you call from the yml file, its value is `/root`. 

Has anyone any idea why this happens and how to fix it?

**Comment by shoogle** [comment]
@cbjeukendrup, all the scripts are being executed with `sudo` for some reason (presumably this solved a permissions problem for @igorkorsukov). Commands run with `sudo` run as the root user, and the root user's home directory is `/root`.

You can try this in the YAML:
```Bash
# replace
sudo bash myscript.sh

# with
sudo HOME="${HOME}" bash myscript.sh
```
Or try solutions in https://stackoverflow.com/questions/8633461/how-to-keep-environment-variables-when-using-sudo

**Comment by cbjeukendrup** [comment]
Haha, exactly at this moment I realized that this must be the problem. Let's see how far we get without `sudo`.

**Comment by shoogle** [comment]
Yes, ideally we wouldn't use `sudo` except where absolutely necessary, and in those cases we should consider using it with `chmod` and `chown` to change permissions before running subsequent commands without `sudo`.

**Comment by cbjeukendrup** [comment]
I'm afraid I'm now really stuck, with the next strange error, when running MuseScore as part of the VTests:
```
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx, xcb.
```
My attempt at aa062df seems to suggest that it's not a permission issue. Maybe some package that came with Ubuntu 18.04 doesn't come with 20.04? I don't really know what to do now.

**Comment by igorkorsukov** [comment]
> Yes, ideally we wouldn't use `sudo` except where absolutely necessary, and in those cases we should consider using it with `chmod` and `chown` to change permissions before running subsequent commands without `sudo`.

Yes, it would be better without use `sudo`, unfortunately I don't remember what problems I had, why I  use `sudo` everywhere.

**Comment by igorkorsukov** [comment]
> I'm afraid I'm now really stuck, with the next strange error, when running MuseScore as part of the VTests:
> 
> ```
> qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
> This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
> Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx, xcb.
> ```
> 
> My attempt at [aa062df](https://github.com/musescore/MuseScore/commit/aa062dfdfd272d35dacd1407e2a8e38bde968b34) seems to suggest that it's not a permission issue. Maybe some package that came with Ubuntu 18.04 doesn't come with 20.04? I don't really know what to do now.

try add `ldd $qt_dir/plugins/platforms/libqxcb.so` here https://github.com/musescore/MuseScore/blob/master/build/ci/linux/setup.sh#L130  
to check that everything is enough to load the plugin

**Comment by igorkorsukov** [comment]
libxcb-randr.so.0 => not found

**Comment by cbjeukendrup** [comment]
Thanks @igorkorsukov! I think it's finally done now! 

(It is expected that `Generate reference PNGs` will fail, because it uses the build script from the master branch, which doesn't include the missing library.)

File: .github/workflows/ci_linux.yml [hunk 1]
```diff
@@ -38,7 +38,7 @@ jobs:
       run: |
         NOW=$(date -u +"%F-%T")
         echo "::set-output name=timestamp::${NOW}"
-    - name: Сcache cache files
+    - name: Ccache cache files
```
[the remark this answers is not in the document]
Comment (shoogle) [reply .github/workflows/ci_linux.yml:41]: The commit message for https://github.com/musescore/MuseScore/pull/12963/commits/919406b45481dd8881dcf0a1e88810a39020a3a8 gives a clue:
> CI: replace Cyrillic С with C

**Comment by cbjeukendrup** [comment]
@igorkorsukov If you're happy with the latest version of this, please merge it :)

(Preferably before Tuesday, when the next "brownout" of 18.04 happens 😄)

**Review by igorkorsukov — APPROVED** [review]

**Pull request merged**